File: python/src/providers/openai/embeddings.py
from typing import List

import httpx

from common.config import Settings, require_openai


def embed_texts(texts: List[str], settings: Settings) -> List[List[float]]:
    require_openai(settings)
    if not texts:
        return []

    url = "https://api.openai.com/v1/embeddings"
    headers = {
        "Authorization": f"Bearer {settings.openai_api_key}",
        "Content-Type": "application/json",
    }
    payload = {
        "model": settings.openai_embed_model,
        "input": texts,
    }

    response = httpx.post(url, headers=headers, json=payload, timeout=settings.timeout_s)
    response.raise_for_status()
    data = response.json()
    try:
        embeddings = [item["embedding"] for item in data["data"]]
    except (KeyError, TypeError) as exc:
        raise RuntimeError("Unexpected response from OpenAI embeddings.") from exc
    return embeddings

File: node/src/providers/openai/embeddings.mjs
import { requireOpenAI } from "../../common/config.js";

const EMBEDDINGS_URL = "https://api.openai.com/v1/embeddings";

/**
 * embedTexts returns embeddings for a batch of texts.
 */
export async function embedTexts(texts, settings) {
  requireOpenAI(settings);
  if (!texts || texts.length === 0) {
    return [];
  }

  const payload = {
    model: settings.openaiEmbedModel,
    input: texts,
  };

  const data = await postJson(EMBEDDINGS_URL, payload, settings);
  const embeddings = data?.data?.map((item) => item.embedding);
  if (!embeddings) {
    throw new Error("Unexpected response from OpenAI embeddings.");
  }
  return embeddings;
}

async function postJson(url, payload, settings) {
  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), settings.timeoutMs);

  try {
    const response = await fetch(url, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${settings.openaiApiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(payload),
      signal: controller.signal,
    });

    const text = await response.text();
    if (!response.ok) {
      throw new Error(`OpenAI request failed (${response.status}): ${text}`);
    }

    return JSON.parse(text);
  } finally {
    clearTimeout(timeoutId);
  }
}